Enable exquisite corpse functionality

zeusttu opened the pull request:
I figured it would be a good idea to make the bot capable of doing this so we can host an exquisite corpse without having someone play game leaser and manually copy-paste everything.

It's a lot of code, I did make an effort to isolate library code into a separate file and keep methods reasonably short and readable, complying with Rubocop's recommendations where possible in the newly introduced library file while staying consistent with the main file's style for my additions to that file.

I also updated the Discord library a bit while troubleshooting, but the old library version ultimately turned out not to be what broke my code, so feel free to revert that if you prefer the old version (though I personally think newer = better most of the time, and the old one uses a deprecated version of Discord's API, so we'll have to upgrade sooner or later anyway I guess.

I haven't been able to test with multiple users, but at least in my own test server (which has just me and my own copy of the bot) it seems to work for as far as it can be tested with just one person.

sarna commented on lib/heimdallr/exquisite_corpse.rb:
```diff
@@ -0,0 +1,230 @@
+        end
+        unless categories.one?
+          names = @server.categories.map(&:name)
+          raise "Exquisit Corpse's channel category 'ec-#{@name}' not found in #{names}."
```
```suggestion
          raise "Exquisite Corpse's channel category 'ec-#{@name}' not found in #{names}."
```
Same for the other exceptions

sarna commented on lib/heimdallr.rb:
```diff
@@ -127,5 +127,59 @@ module Heimdallr
+    ec = ExquisiteCorpse.new event.server, name
+    ec.start participants
+    ec.register bot
+    "Started dxquisite corpse '#{name}'."
```
```suggestion
    "Started exquisite corpse '#{name}'."
```

sarna commented on lib/heimdallr.rb:
```diff
@@ -127,5 +127,59 @@ module Heimdallr
+
+  bot.ready do |event|
+    bot.servers.values.each do |server|
+      ExquisiteCorpse.all(server).values.each { |ec| ec.register bot }
```
This fails for me until I add `require "heimdallr/exquisite_corpse"` at the top of the file

sarna commented on lib/heimdallr.rb:
```diff
@@ -127,5 +127,59 @@ module Heimdallr
+    description: "Start an exquisite corpse.",
+    usage: ",startec <name> <participant1> <participant2> ..."
+  ) do |event|
+    _command, name, *participants = event.content.split
```
Would be nice to fail in case someone forgets the name and/or participants. Right now you can do this:
me: `,startec`
bot: `Started dxquisite corpse ''.`

zeusttu replied:
Changed this, also made it check that there are at least 2 participants.

sarna commented on lib/heimdallr.rb:
```diff
@@ -127,5 +127,59 @@ module Heimdallr
+    "Finalised exquisite corpse '#{name}'."
+  end
+
+  bot.command(
```
I'm getting a some errors from this one:
- if you give it the wrong name, it throws a `NoMethodError` (because the thing is nil)
- if you give it the right name, it doesn't properly remove the corpse and throws exceptions from Discordrb :( Complains about an `Unknown Role`.

zeusttu replied:
I fixed the first error. I also made it ignore missing roles/channels because those are likely to already be deleted, but something's still going wrong when deregistering an exquisite corpse from the internal state:

1. I comment out the check against duplicate participants.
2. I start an exquisite corpse with `,startec foobar @Zeust @Zeust`.
3. I remove it with `,removeec foobar`.
4. Everything gets removed as expected.
5. If I run `,removeec foobar` again, it again claims to have deleted the exquisite corpse. Weird, it's supposed to say "Exquisite corpse not found" here.
6. I start an exquisite corpse with the same name again: `,startec foobar @Zeust @Zeust`.
7. It gets created as expected.
8. I try to delete it again: `,removeec foobar`
9. The roles get properly deleted but the channels don't.

I'll look further into this.

zeusttu replied:
Ahhh, found it, and fixed it!

zeusttu commented:
Sorry for taking so long to respond to feedback here. I kinda lost motivation or got busy I think, don't remember exactly. Got time and motivation now again though, and fixed most of the comments and rebased on your latest code. Still testing and fixing the last few things, I expect to get this done sometime in the coming week.

sarna commented:
No worries. Good that you rebased, old Discordrb versions misbehaved weirdly with current Discord :)
